domain/blockdag/dag.go

```diff
@@ -443,5 +496,182 @@ func (dag *BlockDAG) IsKnownInvalid(hash *daghash.Hash) bool {
 	if !ok {
 		return false
 	}
-	return dag.index.NodeStatus(node).KnownInvalid()
+	return dag.index.BlockNodeStatus(node).KnownInvalid()
+}
+
+func (dag *BlockDAG) addTip(tip *blockNode) (
+	didVirtualParentsChange bool, chainUpdates *selectedParentChainUpdates, err error) {
+
+	newTips := dag.tips.clone()
+	for parent := range tip.parents {
+		newTips.remove(parent)
+	}
+
+	newTips.add(tip)
+
+	return dag.setTips(newTips)
+}
+
+func (dag *BlockDAG) setTips(newTips blockSet) (
+	didVirtualParentsChange bool, chainUpdates *selectedParentChainUpdates, err error) {
+
+	didVirtualParentsChange, chainUpdates, err = dag.updateVirtualParents(newTips, dag.virtual.finalityPoint())
+	if err != nil {
+		return false, nil, err
+	}
+
+	dag.tips = newTips
+
+	return didVirtualParentsChange, chainUpdates, nil
+}
+
+func (dag *BlockDAG) updateVirtualParents(newTips blockSet, finalityPoint *blockNode) (
+	didVirtualParentsChange bool, chainUpdates *selectedParentChainUpdates, err error) {
+
+	var newVirtualParents blockSet
+	// If only genesis is the newTips - we are still initializing the DAG and not all structures required
```

Comment:
What happens is that this function is called in setTips as `dag.updateVirtualParents(newTips, dag.virtual.finalityPoint())`.
However, if we are only inserting genesis rn, then dag.virtual is still not really initialized and dag.virtual.finalityPoint is therefore some bullshit block.

I've invested quite a lot of time to find a better architecture to this whole thing so this could be avoided, but couldn't find anything.
So I'll enhance the comment with what I wrote to you here, and if you have any ideas on how to make it better, please LMK (Or do yourself if I'm already on my vacation).